File: lib/features/knowledge/domain/knowledge_dedup.dart
import '../../../core/utils/text_normalize.dart';
import 'entities/knowledge_note.dart';
import 'entities/knowledge_group.dart';

const _stopwords = {
  'va', 'khi', 'nao', 'cho', 'cua', 'la', 'cac', 'mot', 'voi', 'thi',
  'dung', 'giai', 'thich', 'vi', 'du', 'doi', 'thuong', 'cach', 'nay',
  'the', 'a', 'an', 'and', 'or', 'when', 'how', 'what', 'explain', 'give',
  'example',
};

List<String> _contentTokens(String prompt) {
  final norm = normalizeForSearch(prompt);
  if (norm.isEmpty) return const [];
  return norm
      .split(' ')
      .where((t) => t.isNotEmpty && !_stopwords.contains(t) && t.length > 1)
      .toList();
}

List<String> _phrases(List<String> tokens) {
  final out = <String>[];
  for (var i = 0; i < tokens.length - 1; i++) {
    out.add('${tokens[i]} ${tokens[i + 1]}');
    if (i < tokens.length - 2) {
      out.add('${tokens[i]} ${tokens[i + 1]} ${tokens[i + 2]}');
    }
  }
  return out;
}

List<KnowledgeNote> findRelatedNotes({
  required String prompt,
  required List<KnowledgeNote> notes,
}) {
  final tokens = _contentTokens(prompt);
  if (tokens.isEmpty) return const [];
  final phrases = _phrases(tokens);
  final promptTerms = {...tokens, ...phrases};

  final scored = <({KnowledgeNote note, int score})>[];
  for (final note in notes) {
    final title = normalizeForSearch(note.title);
    final haystack = normalizeForSearch([
      note.title,
      note.summary,
      note.tags.join(' '),
      knowledgeGroupLabel(note.groupId),
    ].join(' '));
    final noteTags = note.tags.map(normalizeForSearch).toSet();

    var score = 0;
    for (final phrase in phrases) {
      if (haystack.contains(phrase)) {
        score += 3;
        if (title.contains(phrase)) score += 2;
      }
    }
    for (final term in promptTerms) {
      if (noteTags.contains(term)) score += 3;
    }
    for (final token in tokens) {
      if (haystack.contains(token)) score += 1;
    }
    if (score >= 3) scored.add((note: note, score: score));
  }

  scored.sort((a, b) => b.score.compareTo(a.score));
  return scored.take(3).map((e) => e.note).toList();
}

File: apps/web/src/lib/knowledgeDedup.ts
// 1:1 port of lib/features/knowledge/domain/knowledge_dedup.dart (post
// whole-branch fix). Layer-1 duplicate check: cheap client-side scoring to
// suggest existing notes related to a new lookup prompt.
import { normalizeForSearch } from "./normalizeSearch";
import { knowledgeGroupLabel } from "./knowledgeGroups";
import type { KnowledgeNote } from "./knowledgeNotes";

// NB: "thi" (thì = tense) and "the" (thể = aspect) are deliberately NOT
// stopwords — they're group labels and the highest-signal dedup terms.
const STOPWORDS = new Set([
  "va", "khi", "nao", "cho", "cua", "la", "cac", "mot", "voi",
  "dung", "giai", "thich", "vi", "du", "doi", "thuong", "cach", "nay",
  "a", "an", "and", "or", "when", "how", "what", "explain", "give", "example",
]);

function contentTokens(prompt: string): string[] {
  const norm = normalizeForSearch(prompt);
  if (!norm) return [];
  return norm.split(" ").filter((t) => t.length > 1 && !STOPWORDS.has(t));
}

function phrases(tokens: string[]): string[] {
  const out: string[] = [];
  for (let i = 0; i < tokens.length - 1; i++) {
    out.push(`${tokens[i]} ${tokens[i + 1]}`);
    if (i < tokens.length - 2) out.push(`${tokens[i]} ${tokens[i + 1]} ${tokens[i + 2]}`);
  }
  return out;
}

export function findRelatedNotes(args: { prompt: string; notes: KnowledgeNote[] }): KnowledgeNote[] {
  const tokens = contentTokens(args.prompt);
  if (tokens.length === 0) return [];
  const ph = phrases(tokens);
  const promptTerms = new Set([...tokens, ...ph]);

  const scored: { note: KnowledgeNote; score: number }[] = [];
  for (const note of args.notes) {
    const title = normalizeForSearch(note.title);
    const haystack = normalizeForSearch(
      [note.title, note.summary, note.tags.join(" "), knowledgeGroupLabel(note.groupId)].join(" "),
    );
    const noteTags = new Set(note.tags.map(normalizeForSearch));

    let score = 0;
    for (const p of ph) {
      if (haystack.includes(p)) {
        score += 3;
        if (title.includes(p)) score += 2;
      }
    }
    for (const term of promptTerms) if (noteTags.has(term)) score += 3;
    for (const t of tokens) if (haystack.includes(t)) score += 1;
    if (score >= 3) scored.push({ note, score });
  }
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, 3).map((s) => s.note);
}
